rfcs/0000-transition-frontier-listener.md

```diff
@@ -0,0 +1,84 @@
+## Summary
+[summary]: #summary
+
+Allow properties based the transition frontier to be efficiently tracked.
+
+## Motivation
+
+[motivation]: #motivation
+
+There are several properties that we might want to obtain from the data in the transition frontier that
+are too expensive to calculate by iterating over all its transitions. However, we don't want to add complexity
+to the transition frontier by having it keep track of these values that are not strictly necessary for its
+operation and are only used externally.
+
+As an example, the snark pool never has elements removed and requires some system of garbage collection.
+One way to do this would be to track a reference count for each piece of `Work` in the table of proofs, but this
+table of references probably shouldn't be part the transition frontier itself as it's completely unrelated.
+
+We'd like to have a way for other parts of the system to incrementally build up datastructures based on the
+breadcrumbs that are being added and removed to the frontier.
+
+Other uses for such an abstraction:
+  - managing the transaction pool
+  - tracking information for consensus optimization
+  - Handling persistence for the transition frontier
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+This design is based on the introduction of a new signature in `Protocols.coda_transition_frontier` as follows.:
+```ocaml
+module type Transition_frontier_listener_intf = sig
+  type t
+  val add_breadcrumb : t -> Breadcrumb.t -> unit;
+  val remove_breadcrumb : t -> Breadcrumb.t -> unit;
+end
+```
+
+The transition frontier would then expose a function for registering listeners. We probably have to do some sort of GADT/ first-class module magic to make this work (???)
+```
+val add_frontier_listener : (module Transition_frontier_listener_intf with type t = 'a) -> 'a -> unit
```

Comment:
I'm not so sure this should be a dynamic action. While we want to be able to easily add new listeners to the transition frontier over time, I don't see a use case for dynamic selection of which listeners to attach. I think we should actually hard code the set of listeners will be attached as a constant, and make a wrapper layer which contains both the transition frontier and all the listeners under unique names in a record. This way, it's easy to consume the listeners and still easy to extend the code to add a new one.